llama_index/chat_engine/types.py

```diff
@@ -42,8 +57,19 @@ class StreamingAgentChatResponse:
     achat_stream: Optional[ChatResponseAsyncGen] = None
     source_nodes: List[NodeWithScore] = field(default_factory=list)
     _queue: queue.Queue = queue.Queue()
-    _is_done = False
+    _aqueue: asyncio.Queue = asyncio.Queue()
+    # flag when chat message is a function call
     _is_function: Optional[bool] = None
+    # flag when processing done
+    _is_done = False
+    # signal when a new item is added to the queue
+    _new_item_event: asyncio.Event = field(default_factory=asyncio.Event)
+    # NOTE: async code uses two events rather than one since it yields
+    # control when waiting for queue item
+    # signal when the OpenAI functions stop executing
+    _is_function_false_event: asyncio.Event = field(default_factory=asyncio.Event)
+    # signal when an OpenAI function is being executed
+    _is_function_not_none_thread_event: Event = field(default_factory=Event)
```

Comment:
this has grown to be quite complicated. 

the culprit seems like the choice we made to expose an iterator while also appending the completed message to history. maybe it's time to reconsider this.

Reply:
Yeah, tried quite hard to simplify it (without arbitrary waits to sync up coroutines/threads) but couldn't see a good way to keep the current behaviour intact and not be tripping over flags and events (hence the comments).

The alternative that springs to mind is just writing to history at the point the `chat_stream` is consumed by the user, rather than spinning up a background thread or coroutine to do this. i.e. we could couple the action of the generator yielding a token with the writing of that token to memory. Is that a reasonable way to go or did you have something else in mind?

Finally, personally I think it would be nicer if `StreamingAgentChatResponse` had a sync and async class. As is, there are essentially two parallel streams of logic that don't intertwine (apart from in `__post_init__`, which could easily be tucked into a base class).

Reply:
Hey, any thoughts here? Still feeling personally like this is the minor thorn in the side of this implementation, but since the PR is quite big already and it has the same behaviour as on main, we could always address it in a dedicated effort.

Reply:
Oh yea, I think @Disiok just meant this to be an observation. At some point in the future we might change the decision to expose a generator, but not sure for now